ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/CommandCompletionsReader.scala

```diff
@@ -8,18 +8,21 @@ import akka.stream.scaladsl.Source
 import com.digitalasset.daml.lf.data.Ref
 import com.digitalasset.ledger.ApplicationId
 import com.digitalasset.ledger.api.v1.command_completion_service.CompletionStreamResponse
-import com.digitalasset.platform.store.CompletionFromTransaction
 
 private[dao] object CommandCompletionsReader {
 
-  // Uses an existing LedgerReadDao to read completions from the ledger entries stream
-  // TODO Replace this to tap directly into the index
-  def apply(reader: LedgerReadDao): CommandCompletionsReader[LedgerDao#LedgerOffset] =
-    (from: Long, to: Long, appId: ApplicationId, parties: Set[Ref.Party]) =>
-      reader
-        .getLedgerEntries(from, to)
-        .map { case (offset, entry) => (offset + 1, entry) }
-        .collect(CompletionFromTransaction(appId, parties))
+  private def offsetFor(response: CompletionStreamResponse): LedgerDao#LedgerOffset =
+    response.checkpoint.get.offset.get.getAbsolute.toLong
+
+  def apply(dispatcher: DbDispatcher): CommandCompletionsReader[LedgerDao#LedgerOffset] =
+    (from: Long, to: Long, appId: ApplicationId, parties: Set[Ref.Party]) => {
+      val query = CommandCompletionsTable.prepareGet(from, to, appId, parties)
+      Source
+        .future(dispatcher.executeSql("get_completions") { implicit connection =>
+          query.as(CommandCompletionsTable.parser.*)
+        })
+        .mapConcat(_.map(response => offsetFor(response) -> response))
```

Comment:
Instead of having to go into the response again for the offset, `CommandCompletionsTable.parser` could return the tuple of offset and response in the first place.

Reply:
I had this in a previous version but `offset` has to be repeated just because it's needed by the `Dispatcher`. Ideally I would only return the response, which is what will happen in a coming PR. But thanks for raising this, it makes sense.